### PR TITLE
more flexible cross section plotting

### PR DESCRIPTION
**Pull request by shimwell**
As requested in issue #2476 this PR adds the ability to plot combinations of multiple nuclides, elements or materials with different cross section reactions.

Previously only one could only plot a single nuclide, element or material.

example script for multiple nuclides
```python
import openmc
import matplotlib.pyplot as plt
fig = openmc.plotter.plot_xs(
    this_and_types = {
        'Li6': [205],
        'Li7': [205],
        'Be9': [16, 205],
    }
)
plt.show()
```

example script for multiple materials
```python
import openmc
import matplotlib.pyplot as plt

mat1 = openmc.Material()
mat1.add_nuclide('Fe56', 1)
mat1.set_density('g/cm3', 1)

mat2 = openmc.Material()
mat2.add_element('Fe', 1)
mat2.add_nuclide('Fe55', 1)
mat2.set_density('g/cm3', 1)
mat2.name = 'Iron'

fig = openmc.plotter.plot_xs(
    this_and_types = {
        mat1: ['total'],
        mat2: ['elastic', 'total'],
    }
)
plt.show()

### CODE REVIEW

**Comment by shimwell** [comment]
@aslakstubsgaard this PR would make multiple material plotting very easy which if I recall correctly you were quite keen on

File: openmc/plotter.py [hunk 1]
```diff
@@ -56,18 +55,61 @@
...
     """Creates a figure of continuous-energy cross sections for this item.
 
     Parameters
     ----------
-    this : str or openmc.Material
-        Object to source data from. Nuclides and elements can be input as a str
-    types : Iterable of values of PLOT_TYPES
-        The type of cross sections to include in the plot.
+
+    this_and_types : dict where keys are str or openmc.Material, values are Iterable of PLOT_TYPES
```
Comment (paulromano) [inline openmc/plotter.py:109]: As long as we're changing the function signature (which I'm ok with -- I don't think this one gets a ton of usage at present), can we come up with a better name than `this_and_types`?

My only hesitance about the change (other than naming issues) is that this makes the simple case (single item/reaction) a bit more complex. But, the added functionality is probably worth it. :man_shrugging:

For documentation purposes, what comes after `:` here needs to be a proper type
```suggestion
    this_and_types : dict
```

**Review by paulromano — CHANGES_REQUESTED** [review]

File: openmc/plotter.py [hunk 2]
```diff
@@ -56,18 +55,61 @@
...
+        msg = "Mixture of openmc.Material and elements/nuclides. Invalid type for plotting"
+        raise TypeError(msg)
+
+    return f'{stem} {mid} {units}'
+
+
+def _get_title(this_and_types):
+    """Gets a title for the type of data plotted"""
+    if len(this_and_types) == 1:
+        this = list(this_and_types.keys())[0]
```
[the remark this answers is not in the document]
Comment (shimwell) [reply openmc/plotter.py:93]: I've never seen that particular way of unpacking return values before. I had to check it out and can confirm this is certainly magic

```python
>>> this_and_types = {'key1':'val1'}
>>> this = list(this_and_types.keys())[0]
>>> this
'key1'
>>> this, = this_and_types
>>> this
'key1'
```

**Comment by shimwell** [comment]
TODO: think of a better arg name than ```this_and_types``` for the dictionary of objects to plot and the reactions to plot

**Comment by shimwell** [comment]
> TODO: think of a better arg name than `this_and_types` for the dictionary of objects to plot and the reactions to plot

Perhaps we can use the old variable name ```this```

**Comment by paulromano** [comment]
@shimwell How about `items`, which at least implies you can have multiple things by virtue of it being plural

File: openmc/plotter.py [hunk 3]
```diff
@@ -56,18 +55,61 @@
...
+    """Gets a label for the element or nuclide or material and reaction plotted"""
+    if isinstance(this, str):
+        return f'{this} {type}'
+    elif this.name is '':
+        return f'Material {this.id} {type}'
+    else:
+        return f'{this.name} {type}'
+
+
+def _get_yaxis_label(this_and_types, divisor_types):
```
Comment (shimwell) [inline openmc/plotter.py:68]:
```suggestion
def _get_yaxis_label(reactions, divisor_types):
```

File: openmc/plotter.py [hunk 4]
```diff
@@ -56,18 +55,61 @@
...
+    """Gets a y axis label for the type of data plotted"""
+
+    if all(isinstance(item, str) for item in this_and_types.keys()):
```
Comment (shimwell) [inline openmc/plotter.py:71]:
```suggestion
    if all(isinstance(item, str) for item in reactions.keys()):
```

File: openmc/plotter.py [hunk 5]
```diff
@@ -56,18 +55,61 @@
...
+        stem = 'Microscopic'
+        if divisor_types:
+            mid, units = 'Data', ''
+        else:
+            mid, units = 'Cross Section', '[b]'
+    elif all(isinstance(item, openmc.Material) for item in this_and_types.keys()):
```
Comment (shimwell) [inline openmc/plotter.py:77]:
```suggestion
    elif all(isinstance(item, openmc.Material) for item in reactions.keys()):
```

File: openmc/plotter.py [hunk 6]
```diff
@@ -56,18 +55,61 @@
...
+        else:
+            mid, units = 'Cross Section', '[1/cm]'
+    else:
+        msg = "Mixture of openmc.Material and elements/nuclides. Invalid type for plotting"
+        raise TypeError(msg)
+
+    return f'{stem} {mid} {units}'
+
+
+def _get_title(this_and_types):
```
Comment (shimwell) [inline openmc/plotter.py:90]:
```suggestion
def _get_title(reactions):
```

File: openmc/plotter.py [hunk 7]
```diff
@@ -56,18 +55,61 @@
...
+    """Gets a title for the type of data plotted"""
+    if len(this_and_types) == 1:
+        this, = this_and_types
```
Comment (shimwell) [inline openmc/plotter.py:93]:
```suggestion
        this, = reactions
```

File: openmc/plotter.py [hunk 8]
```diff
@@ -56,18 +55,61 @@
...
+        name = this.name if isinstance(this, openmc.Material) else this
+        return f'Cross Section Plot For {name}'
+    else:
+        return 'Cross Section Plot'
+
+
+def plot_xs(this_and_types, divisor_types=None, temperature=294., axis=None,
```
Comment (shimwell) [inline openmc/plotter.py:100]:
```suggestion
def plot_xs(reactions, divisor_types=None, temperature=294., axis=None,
```

File: openmc/plotter.py [hunk 9]
```diff
@@ -56,18 +55,61 @@
...
     """Creates a figure of continuous-energy cross sections for this item.
 
     Parameters
     ----------
-    this : str or openmc.Material
-        Object to source data from. Nuclides and elements can be input as a str
-    types : Iterable of values of PLOT_TYPES
-        The type of cross sections to include in the plot.
+
+    this_and_types : dict
```
Comment (shimwell) [inline openmc/plotter.py:109]:
```suggestion
    reactions : dict
```

File: openmc/plotter.py [hunk 10]
```diff
@@ -115,102 +157,86 @@ def plot_xs(this, types, divisor_types=None, temperature=294., axis=None,
...
     # Generate the plot
     if axis is None:
         fig, ax = plt.subplots(**kwargs)
     else:
         fig = None
         ax = axis
+
+    all_types = []
+
+    for this, types in this_and_types.items():
```
Comment (shimwell) [inline openmc/plotter.py:170]:
```suggestion
    for this, types in reactions.items():
```

File: openmc/plotter.py [hunk 11]
```diff
@@ -115,102 +157,86 @@ def plot_xs(this, types, divisor_types=None, temperature=294., axis=None,
...
-            raise TypeError("Invalid type for plotting")
-    ax.set_ylabel(ylabel)
+    ax.set_ylabel(_get_yaxis_label(this_and_types, divisor_types))
     ax.legend(loc='best')
-    name = this.name if isinstance(this, openmc.Material) else this
-    if len(types) > 1:
-        ax.set_title('Cross Sections for ' + name)
-    else:
-        ax.set_title('Cross Section for ' + name)
+    ax.set_title(_get_title(this_and_types))
```
Comment (shimwell) [inline openmc/plotter.py:239]:
```suggestion
    ax.set_title(_get_title(reactions))
```

File: tests/unit_tests/test_plotter.py [hunk 12]
```diff
@@ -74,9 +75,89 @@ def test_calculate_cexs_with_materials(test_mat):
...
 
 def test_plot_xs_mat(test_mat):
     from matplotlib.figure import Figure
-    assert isinstance(openmc.plot_xs(test_mat, types=['total']), Figure)
+    assert isinstance(openmc.plot_xs({test_mat: ['total']}), Figure)
+
+def test_plot_axes_labels():
+    # just nuclides
+    axis_label = openmc.plotter._get_yaxis_label(
+        this_and_types = {
```
Comment (shimwell) [inline tests/unit_tests/test_plotter.py:87]:
```suggestion
        reactions = {
```

File: tests/unit_tests/test_plotter.py [hunk 13]
```diff
@@ -74,9 +75,89 @@ def test_calculate_cexs_with_materials(test_mat):
...
+            'Li6': [205],
+            'Li7': [205],
+        }, divisor_types=False
+    )
+    assert axis_label == 'Microscopic Cross Section [b]'
+
+    # just elements
+    axis_label = openmc.plotter._get_yaxis_label(
+        this_and_types = {
```
Comment (shimwell) [inline tests/unit_tests/test_plotter.py:96]:
```suggestion
        reactions = {
```

File: tests/unit_tests/test_plotter.py [hunk 14]
```diff
@@ -74,9 +75,89 @@ def test_calculate_cexs_with_materials(test_mat):
...
+    # just materials
+    mat1 = openmc.Material()
+    mat1.add_nuclide('Fe56', 1)
+    mat1.set_density('g/cm3', 1)
+    mat2 = openmc.Material()
+    mat2.add_element('Fe', 1)
+    mat2.add_nuclide('Fe55', 1)
+    mat2.set_density('g/cm3', 1)
+    axis_label = openmc.plotter._get_yaxis_label(
+        this_and_types = {
```
Comment (shimwell) [inline tests/unit_tests/test_plotter.py:121]:
```suggestion
        reactions = {
```

File: tests/unit_tests/test_plotter.py [hunk 15]
```diff
@@ -74,9 +75,89 @@ def test_calculate_cexs_with_materials(test_mat):
...
+                this_and_types = {
+                    'Li6': [205],
+                    mat2: [16],
+                }, divisor_types=False
+            )
+
+    # mixed materials and elements
+    with pytest.raises(TypeError):
+        openmc.plotter._get_yaxis_label(
+                this_and_types = {
```
Comment (shimwell) [inline tests/unit_tests/test_plotter.py:140]:
```suggestion
                reactions = {
```

File: tests/unit_tests/test_plotter.py [hunk 16]
```diff
@@ -74,9 +75,89 @@ def test_calculate_cexs_with_materials(test_mat):
...
+                    'Li': [205],
+                    mat2: [16],
+                }, divisor_types=False
+            )
+
+
+def test_get_title():
+    title = openmc.plotter._get_title(this_and_types = {'Li': [205]})
```
Comment (shimwell) [inline tests/unit_tests/test_plotter.py:148]:
```suggestion
    title = openmc.plotter._get_title(reactions = {'Li': [205]})
```

File: tests/unit_tests/test_plotter.py [hunk 17]
```diff
@@ -74,9 +75,89 @@ def test_calculate_cexs_with_materials(test_mat):
...
+        'Li6': [205],
+        'Li7': [205]
+    })
+    assert title == 'Cross Section Plot'
+
+    mat1 = openmc.Material()
+    mat1.add_nuclide('Fe56', 1)
+    mat1.set_density('g/cm3', 1)
+    mat1.name = 'my_mat'
+    title = openmc.plotter._get_title(this_and_types = {mat1: [205]})
```
Comment (shimwell) [inline tests/unit_tests/test_plotter.py:162]:
```suggestion
    title = openmc.plotter._get_title(reactions = {mat1: [205]})
```

File: tests/unit_tests/test_plotter.py [hunk 18]
```diff
@@ -74,9 +75,89 @@ def test_calculate_cexs_with_materials(test_mat):
 @pytest.mark.parametrize("this", ["Be", "Be9"])
 def test_plot_xs(this):
     from matplotlib.figure import Figure
-    assert isinstance(openmc.plot_xs(this, types=['total']), Figure)
-
+    assert isinstance(openmc.plotter.plot_xs({this: ['total', 'elastic']}), Figure)
```
Comment (paulromano) [inline tests/unit_tests/test_plotter.py:78]:
```suggestion
    assert isinstance(openmc.plot_xs({this: ['total', 'elastic']}), Figure)
```

File: tests/unit_tests/test_plotter.py [hunk 19]
```diff
@@ -74,9 +75,89 @@ def test_calculate_cexs_with_materials(test_mat):
...
+    # just materials
+    mat1 = openmc.Material()
+    mat1.add_nuclide('Fe56', 1)
+    mat1.set_density('g/cm3', 1)
+    mat2 = openmc.Material()
+    mat2.add_element('Fe', 1)
+    mat2.add_nuclide('Fe55', 1)
+    mat2.set_density('g/cm3', 1)
+    axis_label = openmc.plotter._get_yaxis_label(
+        reactions = {
```
Comment (paulromano) [inline tests/unit_tests/test_plotter.py:121]:
```suggestion
        reactions={
```

File: tests/unit_tests/test_plotter.py [hunk 20]
```diff
@@ -74,9 +75,89 @@ def test_calculate_cexs_with_materials(test_mat):
...
+        openmc.plotter._get_yaxis_label(
+                reactions = {
+                    'Li': [205],
+                    mat2: [16],
+                }, divisor_types=False
+            )
+
+
+def test_get_title():
+    title = openmc.plotter._get_title(reactions = {'Li': [205]})
```
Comment (paulromano) [inline tests/unit_tests/test_plotter.py:148]:
```suggestion
    title = openmc.plotter._get_title(reactions={'Li': [205]})
```

File: tests/unit_tests/test_plotter.py [hunk 21]
```diff
@@ -74,9 +75,89 @@ def test_calculate_cexs_with_materials(test_mat):
...
+    assert title == 'Cross Section Plot For Li'
+    title = openmc.plotter._get_title(reactions = {'Li6': [205]})
+    assert title == 'Cross Section Plot For Li6'
+    title = openmc.plotter._get_title(reactions = {
```
Comment (paulromano) [inline tests/unit_tests/test_plotter.py:152]:
```suggestion
    title = openmc.plotter._get_title(reactions={
```

File: tests/unit_tests/test_plotter.py [hunk 22]
```diff
@@ -74,9 +75,89 @@ def test_calculate_cexs_with_materials(test_mat):
...
+        'Li6': [205],
+        'Li7': [205]
+    })
+    assert title == 'Cross Section Plot'
+
+    mat1 = openmc.Material()
+    mat1.add_nuclide('Fe56', 1)
+    mat1.set_density('g/cm3', 1)
+    mat1.name = 'my_mat'
+    title = openmc.plotter._get_title(reactions = {mat1: [205]})
```
Comment (paulromano) [inline tests/unit_tests/test_plotter.py:162]:
```suggestion
    title = openmc.plotter._get_title(reactions={mat1: [205]})
```

**Review by paulromano — CHANGES_REQUESTED** [review]
A few little style nitpicks and then this should be good to merge. Thanks!

**Review by paulromano — APPROVED** [review]

**Pull request merged**